File: java/com/lc/lc142.java
package lc;

public class lc142 {
    public ListNode detectCycle(ListNode head) {
        if (head == null) {
            return null;
        }
        ListNode dummy = new ListNode(-1, head);
        ListNode slow = dummy.next, fast = dummy.next.next;
        while (fast != null && fast.next != null && fast != slow) {
            fast = fast.next.next;
            slow = slow.next;
        }
        if (fast != slow || fast == null) {
            return null;
        }
        while (dummy != slow) {
            dummy = dummy.next;
            slow = slow.next;
        }
        return dummy;
    }
}

File: go/lc142.go
package main

type ListNode struct {
	Val  int
	Next *ListNode
}

func detectCycle(head *ListNode) *ListNode {
	fi, se := head, head
	for fi != nil && fi.Next != nil {
		fi = fi.Next.Next
		se = se.Next
		if fi == se {
			for head != se {
				head = head.Next
				se = se.Next
			}
			return head
		}
	}
	return nil
}
